### DISCUSSION TITLE
Binding Unmodified Rust

### DISCUSSION THREAD

**Discussion by coder0xff** [fzyzcjy/flutter_rust_bridge#1116] [Ideas]
The existence of StreamSink, RustOpaque, and more tell us something important about flutter_rust_bridge: It's not an interop layer consuming general Rust functions and presenting them to Dart. It's a way to write Rust functions _for_ Dart. I want to change that. What if flutter_rust_bridge could bind this (pseudo code)?

```rust
pub async fn callback_caller(callback: dyn Fn(Option<Rc<std::fs::File>>) -> std::future::Future<type = ()>) {
    callback(Some(global_file)).await
}
```

```dart
Future<void> callback(StdFsFile? file) {
    ...
}

await api.callback_caller(callback);
```

It would be game-changing. I've started working on architectural changes to flutter_rust_bridge to prepare to move in this direction. I understand if that's not the direction this project wants to go in, but it's something that I've decided is necessary for my use case, and I hope it can become a shared goal.

**Comment by fzyzcjy** [comment]
That looks interesting! So IIRC, the proposal is that, allow Dart functions (callbacks) to be called by Rust functions?

As for architecture changes, it would be great to firstly sketch a draft of what is to be done, then we can discuss about it - otherwise if code is changed in a suboptimal way, the time doing coding may be wasted.

Possibly related: https://github.com/fzyzcjy/flutter_rust_bridge/issues/880, thus also cc @rogurotus 

Also cc @Desdaemon

**Comment by coder0xff** [comment]
The callback is just an example. Any type that can't be serialized but can be handled through other means, should be, transparently. For instance, why write RustOpaque manually when FRB could generate a wrapper function that does it automatically? Make FRB bind everything that it feasible can. (There are things it can't of course.)

**Reply by fzyzcjy** [reply]
I see. Looks interesting!

**Comment by Desdaemon** [comment]
I'm not opposed to changes in this direction, but I feel like certain use-cases may lead to unexpected behavior for the users. For example, to make all unparseable types opaque is probably not ideal and should require an opt-in flag/config. At the very least, we should allow users to inspect the output in a comprehensible format, so that they can raise issues on our side if they think our implementation is inefficient, buggy or outright not supporting a certain type combination.

**Reply by fzyzcjy** [reply]
Agree. Indeed allowing the `callback` feature looks most appealing to me.

**Reply by coder0xff** [reply]
It will have to be done with care certainly. Opt-in is the sensible thing to do.

**Comment by coder0xff** [comment]
@fzyzcjy @Desdaemon (sorry if tagging was unnecessary), here's a sketch as requested.

Dart users don't expect to deal with ownership, lifetimes, and indirection. On the other hand, Rust exposes all of these for function arguments and returns. It's also nonsensical for Dart to borrow (`&` and `&mut`) a non-`'static` Rust object because the borrow checker doesn't work across language boundaries. (It *is* sensible for Rust to borrow a Rust value owned by Dart.) The Dart representation of a Rust object should encapsulate as much of the Rust memory model as possible to simulate the Dart memory model. For aspects that can't or shouldn't be encapsulated, we can leverage Dart constructs like `implements` to abstract them.

### Arguments to Rust Functions

In Rust, we can't call a function with an argument that will take ownership if the caller does not have ownership. This is one example of the Rust memory model that can't be hidden from Dart. Here's the complete list:

* No derivation of ownership from indirection
* No derivation of mut indirection from const indirection
* No derivation of exclusive ownership from shared ownership, or vice-a-versa
* No derivation of Rc from Arc, or vice-a-versa

It's also worth pointing out some of the derivations that can and should be hidden from Dart:

* Derivation of indirection from ownership
* Derivation of const indirection from mut indirection
* Derivation of mut double indirection upon mut or const indirection
* Derivation of const double indirection upon const indirection

To abuse Worth Syntax Notation, I define five equivalence classes:

```
OWNED   = Box<OWNED>
        | T

MUT_IND = *mut T
        | &mut T
        | *mut MUT_IND
        | &mut MUT_IND

IND     = *const T
        | &T
        | *const IND
        | &IND
        | *mut IND
        | &mut IND

RC      = Rc<T>

ARC     = Arc<T>
```

T is any valid Rust type name. Any type constructor that can be derived from the same production is in the same equivalence class. For example, `&mut *const T` and `&T` can be derived from IND and are in the same equivalence class. Type constructors in the same equivalence class are interchangeable to pass an argument to a function. If we have a `&T`, we can invoke a function that wants a `&mut *const T` and vice versa. Generally, it's not possible to move across equivalence classes, except in the following scenarios, which I've dubbed substitutions:

```
OWNED   = IND
        | MUT_IND

RC      = IND
        | MUT_IND

ARC     = IND
        | MUT_IND

MUT_IND = IND
```

To explain by way of example, `Box<T>` is in the OWNED class, and we can compute `&T` in the IND class from it. Therefore, OWNED can be substituted for IND. (By the way, I spent some time investigating OWNED = RC and OWNED = ARC, but there's no way to make fake reference counters without causing memory leaks.)

If T implements Clone, then these additional substitutions are possible:

```
IND     = OWNED
MUT_IND = OWNED
RC      = OWNED
ARC     = OWNED
```

These can probably be piled into the "unexpected behavior" that @Desdaemon mentioned. Dart users expect pass by reference, not pass by value. These types should instead expose a clone method in Dart. I must point out that for values that *can* be serialized, the semantics of flutter_rust_bridge today is to pass them by value. We'll have to think about whether that should change.

Here's a more concrete example of the equivalence classes:

#### dump_file version 1

```Rust
/// dump a File, taking ownership and destroying the File when done
pub fn dump_file(file: std::fs::File) {
    println!("{}", file.read_to_string());
}
```

#### dump_file version 2

```Rust
/// dump a Box<File>, taking ownership and destroying the File when done
pub fn dump_file(file: Box<std::fs::File>) {
    println!("{}", file.read_to_string());
}
```

Dart needn't care about whether the argument is `Box`ed. All it needs to know is it owns a `File`, it will lose ownership when it calls `dump_file`, and the interop layer will take care of the rest. (Maybe !Sized: T needs to be its own class too) One final wrinkle: what if the Dart user *wants* to interact directly with pointers or reference counters? I said that `T` above is a type name. The good news is that we could tuck a pointer or ref counter into that `T`, and the rules still work, so long as the "no derivation of mut indirection from const indirection" rule isn't violated. If this capability is added, it would be follow-up work.

#### Equivalence Class Abstraction

In the same way FrbOpaque is inherited to make Rust-type-specific instantiations, five new Dart classes will be introduced that will be inherited to make Rust-type-specific instantiations. (Possible ten if we want "NoSend" variants for T: !Send. (12 for !Sized, maybe.))

| Dart Base Type | Dart Derived Type Generated for Rust Type `MyType` | Notes |
| --- | --- | --- |
| FrbInd | FrbIndMyType | `fn f(&self, ...)` method bindings |
| FrbMutInd | FrbMutIndMyType | `implements FrbIndMyType`, `fn f(&mut self, ...)` method bindings |
| FrbOwned | FrbOwnedMyType | `implements FrbMutIndMyType`, `fn f(self, ...)` method bindings |
| FrbRc | FrbRcMyType | `implements FrbMutIndMyType` |
| FrbArc | FrbArcMyType | `implements FrbMutIndMyType` |

As in the `RustOpaque` example, the `var` keyword keeps the type name from the user. (I *am* concerned that the different collections of method bindings (see Notes above) could confuse the user if the type name is hidden. "Why can't I use this method?" Maybe we can abuse Darts @Deprecated annotation to notify the user?) The `implements` relations are modeled after the aforementioned substitution rules. They enable Dart bindings of Rust functions to parameterize with the least derived type but be called with the more derived type due to Liskov Substitution Principle (LSP). For example:

```Dart
void do_a_thing(FrbIndirectionMyType value) { /* send to Rust */ }
FrbOwnedMyType make_a_value() { /* get from Rust */ }

var value = make_a_value();
do_a_thing(value); // OK, because FrbOwnedMyType implements FrbIndirectionMyType
```

It's also theoretically possible for a binding to return, for example, a FrbOwnedMyType, despite having a declared return type of FrbIndirectionMyType due to the LSP, but that's probably not useful in practice and diverges from the Rust memory model anyway.

### Return Values from Rust Functions

The equivalence classes are helpful here as well, with a tweak. Dart cannot generally borrow, except for `&'static` and `&'static mut`, so replace all occurrences of `&` with `&'static `. To send a return value to Dart, in order:
- Convert to the first variant of the equivalence class. E.g., "&'static mut T` becomes `*mut T`, and `T` becomes `Box<T>`.
- Serialize and transfer to Dart, including ownership, based on the type constructor:
  - *const T - serialize into FrbInd
  - *mut T - serialize into FrbMutInd
  - Box<T> - `Box::into_raw` and serialize into FrbOwned
  - Arc<T> - If Arc::try_unwrap works then change to Box<T>, otherwise mem::forget, serialize *const Arc<T> into FrbArc (like FrbOpaque)
  - Rc<T> - If Rc::try_unwrap works then change to Box<T>, otherwise mem::forget, serialize *const Rc<T> + thread::ThreadID into FrbRc

Again, if T implements Clone, it's feasible to work with borrowed return values, but the same pass-by-value concerns apply.

### Rust-Function Valued Returns from Rust Functions

The feasibility of this was shown in https://github.com/fzyzcjy/flutter_rust_bridge/discussions/986. On the Dart side, it'll be an object that implements invocation, forwarding arguments to the Rust function.

### Dart-Function Valued Arguments to Rust Functions
Similar conceptually to Rust callback functions but requires a bit more.

```Rust
pub fn callback_caller(callback: Fn(String) -> i64) {
    assert(callback("1337") == 1337);
}

```Dart
int dartCallback(String str) {
  return int.tryParse(str);
}
```

This can work similarly to StreamSink with the addition of multiplexing. On the Rust side, the Fn is made from a closure. If this is done right, we can even let the user hang themselves with callback hell. xD

### Marshaling

Marshaling exposes the members of an unserializable Rust object to Dart. For example, a Rust `std::fs::File` object cannot be serialized. Nonetheless, a Dart class like the following is possible:

```Dart
class StdFsFile {
  final Platform _platform;
  final RustIndirection _native;

  int write(UInt8List buf) {
    _platform.inner.std_fs_file_write(_native, buf);
  }
  ...  
}
```

### It's Getting Late

This is pretty long already, so at the very least, it's time for a first round of discussion, especially since I'm still new to Rust and flutter_rust_bridge (how many months can I keep saying that?). Please let me know what you think, if I need to add some details, if something is wrong, etc. Cheers!

**Reply by fzyzcjy** [reply]
Oh that's a long and detailed proposal - good job! I will check a bit later

IMHO discussions are unordered, thus maybe it is more suitable to create an issue and discuss there? :)

**Reply by fzyzcjy** [reply]
The proposal is exciting - especially, I love how we can call Dart callbacks from Rust and vise versa! IMHO this is quite useful, because `closure` (or "function as first-class citizen") is a commonly used language feature for modern languages.

To be honest, I am a little bit confused why we need FrbInd/FrbMutInd/..., etc. We know frb currently passes (copy or move) everything by value (except for Opaque) between Dart and Rust. Therefore, for example, no matter what Rust semantics say (e.g. owned, indirection, ...), we still create a brand new object (or move - if you are using ZeroCopyBuffer) in the Dart side. Thus, I am not very sure the use of FrbInd/FrbMutInd/...

**Reply by fzyzcjy** [reply]
**For future readers:** The discussion is continued on https://github.com/fzyzcjy/flutter_rust_bridge/issues/1132
